### PR TITLE
feat(openid-connect): expose unauth_action parameter in lua-resty-openidc

### PR DESCRIPTION
**Pull request by jnystad**
### Description

Allow users to specify the unauthenticated action in openid-connect plugin.

This is useful when APIs should not be called directly by users, but respond with 401 directly (deny), or when endpoints should validate existing sessions but still allow anonymous requests (pass).

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by tzssangglass** [comment]
hi @jnystad would you continue to work on this PR?

**Comment by jnystad** [comment]
Yes, I'm currently testing this in our development cluster a bit to ensure it works appropriately. Will update this PR when it is verified.

**Comment by jnystad** [comment]
I think this is ready for review. I can't get the test harnesses running in my Windows environment. Would adding tests for this feature be required?

I also see some CI checks failing, but it seems unrelated to this change. Could you confirm?

**Comment by tzssangglass** [comment]
> Would adding tests for this feature be required?

yes, we need to add test cases to cover this.

> I also see some CI checks failing, but it seems unrelated to this change. Could you confirm?

I checked and confirmed that it was related to your commit, as the error was running `t/plugin/openid-connect.t`. ref: https://github.com/apache/apisix/actions/runs/3385510161/jobs/5634897421

**Comment by jnystad** [comment]
Made an attempt at test cases, but I have to rely on your CI process to verify this. Are you OK with that?

**Comment by moonming** [comment]
@tzssangglass do you have the plan to help fix test cases of this PR?

**Comment by tzssangglass** [comment]
> @tzssangglass do you have the plan to help fix test cases of this PR?

Not at the moment, I'm not familiar enough with the openid-connect plugin.

**Comment by moonming** [comment]
> > @tzssangglass do you have the plan to help fix test cases of this PR?
> 
> Not at the moment, I'm not familiar enough with the openid-connect plugin.

@starsz please take a look.

**Comment by starsz** [comment]
> > > @tzssangglass do you have the plan to help fix test cases of this PR?
> > 
> > 
> > Not at the moment, I'm not familiar enough with the openid-connect plugin.
> 
> @starsz please take a look.

OK.I can help @jnystad fix the test cases next week.

**Comment by starsz** [comment]
> > > > @tzssangglass do you have the plan to help fix test cases of this PR?
> > > 
> > > 
> > > Not at the moment, I'm not familiar enough with the openid-connect plugin.
> > 
> > 
> > @starsz please take a look.
> 
> OK.I can help @jnystad fix the test cases in two weeks.

Hi, @jnystad .
Can you add permission to the repo that you cloned from Apache APISIX ?

Or you can change as the follow to fix the CI.

```
diff --git a/t/plugin/openid-connect2.t b/t/plugin/openid-connect2.t
index 6b484c5b..40f6cadb 100644
--- a/t/plugin/openid-connect2.t
+++ b/t/plugin/openid-connect2.t
@@ -260,7 +260,7 @@ passed
             local httpc = http.new()
             local uri = "http://127.0.0.1:" .. ngx.var.server_port .. "/hello"
             local res, err = httpc:request_uri(uri, {method = "GET"})
-            if ngx.status == 200 then
+            if res.status == 200 then
                 ngx.say(true)
             end
         }
```

**Comment by jnystad** [comment]
Thanks! Pushed the change.

**Comment by starsz** [comment]
> Thanks! Pushed the change.

Oh. Maybe you should merge the master to fix the conflicts.

File: apisix/plugins/openid-connect.lua [hunk 1]
```diff
@@ -328,15 +336,26 @@ function _M.rewrite(plugin_conf, ctx)
         -- Either token validation via introspection endpoint or public key is
         -- not configured, and/or token could not be extracted from the request.
 
+        local unauth_action = conf.unauth_action
+        if unauth_action ~= "auth" then
+            unauth_action = "deny"
+        end
```
Comment (starsz) [inline apisix/plugins/openid-connect.lua:342]: It looks like if `unauth_action` is "pass", it will be "deny" .

Comment (jnystad) [reply apisix/plugins/openid-connect.lua:342]: "pass" is handled in the plugin, not in the oidc library. See line 354

Comment (starsz) [reply apisix/plugins/openid-connect.lua:342]: Get it.

**Review by starsz — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**